### ISSUE TITLE
Services ignored because of service description (refferring issue #1676)

### ISSUE THREAD

**Issue by fullmetalucard** [naparuba/shinken#1705]
Hello,

First of all thank you for your answers and work, but as @Pi3R1k said in issue #1676 , part of the problem is still there. That's why I open a new issue.
We reinstalled a fresh test platform  and applied your patch and sample configuration with the sys: $KEY$. The problem is solved only for the case you mentionned:

define service{
        service_description     sys: $KEY$
        use                     gen-service-5,gen-pnp
        host_name               T-Linux
        check_command           check_nrpe_args!check_cpu!$_HOST_CPU_WARN$!$_HOST_CPU_CRIT$
        register 0
        duplicate_foreach       __CPU
        __CPU_WARN                    85
    __CPU_CRIT                    90
}

define service{
        service_description     sys: $KEY$
        use                     gen-service-15,gen-pnp
        host_name               T-Linux
        check_command           check_nrpe_args!check_disk!$KEY$!$_HOST_DISK_WARN$!$_HOST_DISK_CRIT$
        duplicate_foreach       __DISK
        __DISK_WARN                     10%
        __DISK_CRIT                     5%
        register 0
}

With that configuration we effectively get Sys: CPU1,  Sys: /var ( all macros are expanded) for CPU and disk. But in your case you made the test with something written before the "$KEY$. I really insist on that point.

In the examples given in my first post or given by @jthebert  you should be able to reproduce our issue.
Here are the full details if you want to reproduce with the vmware script; I added the command definition.

Ex:

This service is not scheduled on our platform ( only "$KEY$" is present in the service description )

define service{
    host_name                 vmware
    name                      generic-esx3_host_vm_list
    use                       generic-service
    service_description       $KEY$
    register                  0
    check_command             check_esx3_host_vm_list!$KEY$
    check_interval        5
    duplicate_foreach         __VMWARE_ESX_HOST_VM_LIST
}

Then we declare an host using the vmware template

define host{
           use                                 client_serveurs,vmware
        host_name                           ESXI01
        address                             XXXXXXX
        realm                               realm_client
        __VMWARE_ESX_STATUS                 Status
        __VMWARE_ESX_HOST_VM_LIST           VM list
}

In the command definition:

define command {
    command_name    check_esx3_host_vm_list
    poller_tag       POLLERVMWARE_DFI
    command_line    /etc/shinken/conf.d//appliance/vmware/check_esx3_api.pl  -H $HOSTADDRESS$ \
                                                                -u $_HOST_VMWARE_USER$ \
                                                                -p $_HOST_VMWARE_PASSWORD$ \
                                                                -l runtime \
                                                                -s list
}

I admit this method is a trick to use the duplicate_foreach function  in every check definition, (even if we don't have to really use the macro expansion in that kind of configuration) so that we can have very small definitions/calls in the hosts configuration files. It's very usefull.

I hope I'm clear enough, so that you can reproduce the bug with only "$KEY$" defined in the service description.

HTH

Regards,

**Comment by olivierHa** [comment]
Hello,

Could you make a test case with a non working example ?

Olivier

2015-08-27 10:42 GMT+02:00 fullmetalucard <notifications@github.com>:

**Comment by fullmetalucard** [comment]
Ok,

I found out something new.

We can start with the bad_inh.cfg you put in the test folder of the master branch:
The config below works with the service affected directly to the host without intermediate template:
```
define host{
        use   generic-host
        host_name MYHOST
        __NRPE_MEMORY           mem$(10%!5%)$
        __NRPE_STD_UPTIME     Uptime$(20)$
}

define host{
        use   generic-host
        host_name MYHOST2
        __NRPE_MEMORY            mem$(10%!5%)$
        __NRPE_STD_UPTIME               Uptime$(20)$
}

define service {
    host_name               MYHOST
    service_description      $KEY$
    check_command           check_nrpe!linux-memory!$VALUE$
    duplicate_foreach       __NRPE_MEMORY
    use                     generic-service,graph
}

define service{
    host_name               MYHOST
    service_description     $KEY$
    use                     generic-service,graph
    check_command           check_nrpe!check_uptime!$VALUE$
    duplicate_foreach      __NRPE_STD_UPTIME
}
define service {
    host_name               MYHOST2
    service_description      $KEY$
    check_command           check_nrpe!linux-memory!$VALUE$
    duplicate_foreach       __NRPE_MEMORY
    use                     generic-service,graph
}

```

Now with the config below using the intermediate template "linux2", it doesn't work:
```
## Intermediate template
define host {
        name             linux2
        use              generic-host
        register         0
        hostgroups       +linux
}
# Hosts
define host{
        use   generic-host,linux2
        host_name MYHOST
        __NRPE_MEMORY           mem$(10%!5%)$
        __NRPE_STD_UPTIPME      Uptime$(20)$
}

define host{
        use   generic-host,linux2
        host_name MYHOST2
        __NRPE_MEMORY           mem$(10%!5%)$
        __NRPE_STD_UPTIPME      Uptime$(20)$
}
#Services
define service {
    host_name               linux2
    name                    linux-nrpe-memory
    service_description     $KEY$
    check_command           check_nrpe!linux-memory!$VALUE$
    duplicate_foreach       __NRPE_MEMORY
    use                     generic-service,graph
    icon_image              services/memory.png
    register                0
}
define service{
    host_name               linux2
    name                    linux-nrpe_std-uptime
    service_description     $KEY$
    use                     generic-service,graph
    check_command           check_nrpe!check_uptime!$VALUE$
    duplicate_foreach      __NRPE_STD_UPTIPME
    register                0
}
```
With this configuration, only the last check(Uptime) is chown in the webui.
If I modify one of the two services with a prefix before "$KEY$" in the service_description section, it works:
```
define service {
    host_name               linux2
    name                    linux-nrpe-memory
    service_description    test: $KEY$
    check_command           check_nrpe!linux-memory!$VALUE$
    duplicate_foreach       __NRPE_MEMORY
    use                     generic-service,graph
    icon_image              services/memory.png
    register                0
}
define service{
    host_name               linux2
    name                    linux-nrpe_std-uptime
    service_description     $KEY$
    use                     generic-service,graph
    check_command           check_nrpe!check_uptime!$VALUE$
    duplicate_foreach      __NRPE_STD_UPTIPME
    register                0
}
```
Now you should be able to reproduce it with this configuration.
So The question is why does  it work without intermediate template and not through it?

Regards,

**Comment by geektophe** [comment]
Got it.

Services are indexed internally (for performance purpose) using a combination of both `host_name` and `service_description` (true for both items and templates).

In your second example, both your templates are indexed under the exact same name `linux2/$KEY$` because `$KEY$` is only expanded later in the process.

I'll have a look at that and see if there's a solution.

**Comment by geektophe** [comment]
I found where the problem came from. It's fixed by https://github.com/naparuba/shinken/pull/1708

**Comment by fullmetalucard** [comment]
That's really good news! I can confirm it does work since we've patched the service.py file. The way doublons were detected has been modified since the previous versions (my guess is since 2.2). It seems that when we used to duplicate services in templates for specific purpose, we never changed the "name" section and shinken used to tolerate this behavior, but not anymore.
Then after patching we still had a few (about one thousand..) checks missing, but we made some corrections in our templates to avoid to have several services with a same name description; which (i totally agree) is normal behavior. The service_description used for the display in the webui and for the generation of the PNP graphs also stayed the same so everything's fine.

Many thanks for the fix!
